projecttacoma/deqm-test-server
# Static export

# Summary
[import-pnp.md](https://github.com/smart-on-fhir/bulk-import/blob/master/import-pnp.md#parameters) describes how a FHIR Parameteres resource MAY include a `exportType` parameter with a string of `static` or `dynamic`. This document also explains the difference is how the two types are handled. Before, we only handled if the `exportType` was `dynamic` because if `exportType` is omitted, it defaults to `dynamic`. Now, the `static` `exportType` is implemented.

## New behavior
A new function is added to retrieve `exportType` from a FHIR parameters resource if it exists. If it does, then that is used in `executePingAndPull`.

## Code changes
- `src/server/importWorker.js` - `exportType` was added as a function parameter.
- `src/services/import.service.js` - `exportType` was added as a function parameter.
- `src/service/measure.service.js` - `exportType` was added as a function parameter.
- `src/util/exportUtils.js` - `retrieveExportType` is a new function that gets the `exportType` from a FHIR parameters resource if it exists. If it does not exist, then it defaults to "dynamic".
- `src/util/operationValidationUtils.js` - delete `checkExportType` error handling function since we now support static export.

# Testing guidance
- This goes hand-in-hand with the PR in the `bulk-data-utilities` repository, so make sure you read both of them as well as the [Jira ticket ](https://jira.mitre.org/browse/TACOSTRAT-1126). 
- After running `npm link` in the `bulk-data-utilities` repo, run `npm link bulk-data-utilities` in this repo.
- Run `npm start` in this repo. 
- Run `PORT=3001 npm start` in the `bulk-export-server` repo.
- Open insomnia and set up the following POST request: 

![Screen Shot 2022-10-05 at 11 28 37 AM](https://user-images.githubusercontent.com/30158156/194100282-a8e4f7c3-08c3-43d8-aa8d-6e7dfb5e9136.png)

Make sure it has the following headers:
![Screen Shot 2022-10-05 at 11 28 24 AM](https://user-images.githubusercontent.com/30158156/194100352-1a0d8262-5db5-426f-93e6-a0e2671a6ef4.png)

- Hit send and make sure the response is 202 Accepted. Look at the outputs in `deqm-test-server` that show this hasn't failed. 
- Look at the header of the response and copy the Content-Location link. Send a GET request to that link and make sure that the status is 200 OK.
- Change the POST request to the following (with the same headers as before):

<img width="745" alt="Screen Shot 2022-10-05 at 11 35 47 AM" src="https://user-images.githubusercontent.com/30158156/194102509-3e3fcd2d-7562-493f-a8da-6d2f132450ba.png">

- Do the same steps as above. If you want to see what happened before the static export was implemented, unlink the bulk-data-utilities and do these steps in the main branch of this repo. There used to be an error thrown but now there is not.

- Look through the code and let me know if you need any help!
